definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/styled-components_v2.x.x.js

```diff
@@ -0,0 +1,369 @@
+  polygon:                  $npm$styledComponents$StyledComponentsComponentListValue,
+  polyline:                 $npm$styledComponents$StyledComponentsComponentListValue,
+  radialGradient:           $npm$styledComponents$StyledComponentsComponentListValue,
+  rect:                     $npm$styledComponents$StyledComponentsComponentListValue,
+  stop:                     $npm$styledComponents$StyledComponentsComponentListValue,
+  svg:                      $npm$styledComponents$StyledComponentsComponentListValue,
+  text:                     $npm$styledComponents$StyledComponentsComponentListValue,
+  tspan:                    $npm$styledComponents$StyledComponentsComponentListValue,
+|}
+
+declare module 'styled-components' {
+  declare type Interpolation                                              = $npm$styledComponents$Interpolation;
+  declare type NameGenerator                                              = $npm$styledComponents$NameGenerator;
+  declare type Theme                                                      = $npm$styledComponents$Theme;
+  declare type ThemeProviderProps                                         = $npm$styledComponents$ThemeProviderProps;
+  declare type TaggedTemplateLiteral<R>                                   = $npm$styledComponents$TaggedTemplateLiteral<R>;
+
+  declare type ReactComponentFunctional<Props: {}, DefaultProps: ?{} = *> = $npm$styledComponents$ReactComponentFunctional<Props, DefaultProps>;
+  declare type ReactComponentFunctionalUndefinedDefaultProps<Props: {}>   = $npm$styledComponents$ReactComponentFunctionalUndefinedDefaultProps<Props>;
+  declare type ReactComponentClass<Props: {}, DefaultProps: ?{} = *>      = $npm$styledComponents$ReactComponentClass<Props, DefaultProps>;
+  declare type ReactComponentClassUndefinedDefaultProps<Props: {}>        = $npm$styledComponents$ReactComponentClassUndefinedDefaultProps<Props>;
+  declare type ReactComponentUnion<Props>                                 = $npm$styledComponents$ReactComponentUnion<Props>;
+  declare type ReactComponentIntersection<Props>                          = $npm$styledComponents$ReactComponentIntersection<Props>;
+  declare type ReactComponentStyledStaticProps<Props>                     = $npm$styledComponents$ReactComponentStyledStaticProps<Props, $Subtype<$Keys<$npm$styledComponents$StyledComponentsComponentList<*>>>>;
+  declare type ReactComponentStyled<Props>                                = $npm$styledComponents$ReactComponentStyled<Props, $Subtype<$Keys<$npm$styledComponents$StyledComponentsComponentList<*>>>>;
+  declare type ReactComponentStyledTaggedTemplateLiteral<Props>           = $npm$styledComponents$ReactComponentStyledTaggedTemplateLiteral<Props, $Subtype<$Keys<$npm$styledComponents$StyledComponentsComponentList<*>>>>;
+
+  declare module.exports: {
+    <Props>(ReactComponentUnion<Props>): ReactComponentStyledTaggedTemplateLiteral<Props>,
```

Comment:
This should also be able to take a string.

Reply:
What do you mean?

Reply:
Just saw you edited your comment. Can you elaborate on "sounds brittle and annoying"?

Reply:
> What do you mean?

`styled('div')` is valid.  That's where [`styled.div` comes from](https://github.com/styled-components/styled-components/blob/master/src/constructors/styled.js#L9-L11), after all.

> Can you elaborate on "sounds brittle and annoying"?

Firstly, it'd have to be updated whenever a new tag is added to the spec.  Far from a daily thing, but still.
Secondly, it'd make factory functions a pain.
```
function makeThing(tag: string) {
  return styled(tag)`/* styles */`;
}
```
wouldn't work, since `tag` isn't a `$Keys<$npm$styledComponents$StyledComponentsComponentList>`.

Reply:
Also, it wouldn't work with custom elements.  I don't know why you would use custom elements with React, but I'm sure somebody would complain about it.

Reply:
Ah, you're right about it needing to take a ``string``. I'll update that.

Reply:
I originally added ``Keys<$npm$styledComponents$StyledComponentsComponentList>`` so that you get really helpful typeahead suggestions, but of course that's not relevant to passing in a string.

Reply:
What do you think about exporting the ``$npm$styledComponents$StyledComponentsComponentListKeys`` type for factory functions?

Reply:
This is unsafe 99% of the time... Keeping this library definition up-to-date should be trivial, as new native elements are extremely infrequent.
```js
function makeThing(tag: string) {
  return styled(tag)`/* styles */`;
}
```

Reply:
Exporting `StyledComponentsComponentListKeys` would work, yeah.